// @pichxyaponn/tw-angpao/dist/cjs/error.class.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/error.class.ts
var error_class_exports = {};
__export(error_class_exports, {
  ApiError: () => ApiError,
  JsonParseError: () => JsonParseError,
  NetworkError: () => NetworkError,
  ValidationError: () => ValidationError
});
module.exports = __toCommonJS(error_class_exports);
var ValidationError = class extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
    this.name = "ValidationError";
  }
};
var NetworkError = class extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
    this.name = "NetworkError";
  }
};
var ApiError = class extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
    this.name = "ApiError";
  }
};
var JsonParseError = class extends Error {
  constructor(message, originalError) {
    super(message);
    this.code = "INVALID_JSON_RESPONSE";
    this.name = "JsonParseError";
    this.cause = originalError;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApiError,
  JsonParseError,
  NetworkError,
  ValidationError
});
